src/S3/S3Transfer/MultipartUploader.php

```diff
@@ -0,0 +1,276 @@
+<?php
+namespace Aws\S3\S3Transfer;
+
+use Aws\CommandPool;
+use Aws\HashingStream;
+use Aws\PhpHash;
+use Aws\ResultInterface;
+use Aws\S3\S3ClientInterface;
+use Aws\S3\S3Transfer\Models\UploadResult;
+use Aws\S3\S3Transfer\Progress\TransferListenerNotifier;
+use Aws\S3\S3Transfer\Progress\TransferProgressSnapshot;
+use GuzzleHttp\Promise\Create;
+use GuzzleHttp\Promise\PromiseInterface;
+use GuzzleHttp\Psr7\LazyOpenStream;
+use GuzzleHttp\Psr7\Utils;
+use HashContext;
+use Psr\Http\Message\StreamInterface;
+use Throwable;
+
+/**
+ * Multipart uploader implementation.
+ */
+class MultipartUploader extends AbstractMultipartUploader
+{
+
+    /** @var int */
+    protected int $calculatedObjectSize;
+
+    /** @var StreamInterface */
+    private StreamInterface $body;
+
+    /** @var HashContext */
+    private HashContext $hashContext;
+
+    public function __construct(
+        S3ClientInterface $s3Client,
+        array $putObjectRequestArgs,
```

Comment:
question: since the multipart (upload) workflow only uses `CreateMultipartUpload`, `UploadPart` and `CompleteMultipartUpload`, should we call this something else? `putObject` is only used for a single upload